File: finance/tvm/annuities/annuities.py
class Annuity:
    def init(self):
        pass
class OrdAnnuity(Annuity):
    @staticmethod
    def pV(pmt,rate,time):
        pv = pmt * ( (1 - ( (1 + rate)**(-time) )) / rate )
        format_ans = f"The Present value of {pmt} is {pv} when discounted at rate {rate} for {time} time in Ordinary Annuity"
        return pv,format_ans
    @staticmethod
    def fV(pmt,rate,time):
        fv = pmt * (( ((1 + r)**t) - 1 ) / 1 ) 
        format_ans = f"The Future value of {pmt} is {fv} at {rate} rate in {time} time in Ordinary Annuity"
        return fv,format_ans
class DueAnnuity(OrdAnnuity):
    @staticmethod
    def pV(pmt,rate,time):
        ord_ann, _ = OrdAnnuity.pV(pmt,rate,time)
        pv = ord_ann * (1 + rate)
        format_ans =  format_ans = f"The Present value of {pmt} is {pv} when discounted at rate {rate} for {time} time in Annuity Due"
        return pv,format_ans
    @staticmethod
    def fV(pmt,rate,time):
        ord_ann, _ = OrdAnnuity.fV(pmt,rate,time)
        pv = ord_ann * (1 + rate) 
        ormat_ans = f"The Future value of {pmt} is {fv} at {rate} rate in {time} time in Annuity Due"
        return fv, format_ans
        
pv_ord,ff = OrdAnnuity.pV(1000,0.05,10)
print(f"{pv_ord} & {ff}")

File: finance/tvm/annuities/annuities.cpp
#include "utils.hpp"

#include <iostream>
#include <string>
#include <cmath>
#include <utility>
#include <cstdint>
#include <sstream>

class Annuity{
public:
    virtual ~Annuity() = default;
};

class OrdAnnuity : public Annuity {
public:
    static std::pair<std::int64_t,std::string> pV(std::string pmt,std::string rate,std::string time) {
        std::int64_t pv = utils::turn_to_ll(pmt) * (1 - ( std::pow((1 + utils::turn_to_ll(rate) ), utils::turn_to_ll(time)) ) / utils::turn_to_ll(rate));
        std::ostringstream format_ans; 
        format_ans << "The present value of " << pmt << " is " << pv << " when discounted at rate "
            << rate << " for time " << time << " on Ordinary Annuity";
        return {pv,format_ans.str()};
    }
    static std::pair<std::int64_t, std::string> fV(std::string pmt,std::string rate,std::string time) {
        std::int64_t fv = utils::turn_to_ll(pmt) * ( (std::pow((1 + utils::turn_to_ll(rate)), utils::turn_to_ll(time))) / utils::turn_to_ll(rate) );
        std::ostringstream format_ans;
        format_ans << "The future value of " << pmt << " is " << fv << " when compounded at rate "
            << rate << " for time " << time << " on Ordinary Annuity";
        return {fv,format_ans.str()};
    }
};

class DueAnnuity : public Annuity{
public:
    static std::pair<std::int64_t, std::string> pV(std::string pmt,std::string rate,std::string time) {
        auto [ord_pv, _ ] = OrdAnnuity::pV(pmt,rate,time); 
        std::int64_t pv = ord_pv * (1 + utils::turn_to_ll(rate));
        std::ostringstream format_ans;
        format_ans << "The present value of " << pmt << " is " << pv << " when discounted at rate "
            << rate << " for time " << time << " on Annuity Due";
        return {pv,format_ans.str()};
    }
    static std::pair<std::int64_t,std::string> fV(std::string pmt,std::string rate,std::string time) {
        auto [ord_fv, _ ] = OrdAnnuity::fV(pmt,rate,time);
        std::int64_t fv = ord_fv * (1 + utils::turn_to_ll(rate));
         std::ostringstream format_ans;
        format_ans << "The future value of " << pmt << " is " << fv << " when componuded at rate "
            << rate << " for time " << time << " on Annuity Due";
        return {fv,format_ans.str()};     
    }
};   
